tests/qos/files/mellanox/qos_param_generator.py

```diff
@@ -0,0 +1,150 @@
+        self.qos_parameters['pkts_num_trig_egr_drp'] = pkts_num_trig_egr_drp
+        self.qos_parameters['pkts_num_hysteresis'] = hysteresis
+
+    def calculate_parameters(self):
+        """
+            Generate qos test parameters based on the configuration
+                xon
+                xoff
+                wm_pg_headroom
+                wm_pg_shared_lossless
+                wm_q_shared_lossless
+                lossy_queue_1
+                wm_pg_shared_lossy
+                wm_q_shared_lossy
+                wm_buf_pool_lossless
+                wm_buf_pool_lossy
+        """
+        pkts_num_trig_pfc = self.qos_parameters['pkts_num_trig_pfc']
+        pkts_num_trig_ingr_drp = self.qos_parameters['pkts_num_trig_ingr_drp']
+        pkts_num_dismiss_pfc = self.qos_parameters['pkts_num_dismiss_pfc']
+        pkts_num_trig_egr_drp = self.qos_parameters['pkts_num_trig_egr_drp']
+        pkts_num_hysteresis = self.qos_parameters['pkts_num_hysteresis']
+
+        xoff = {}
+        xoff['pkts_num_trig_pfc'] = pkts_num_trig_pfc
+        xoff['pkts_num_trig_ingr_drp'] = pkts_num_trig_ingr_drp
+        # One motivation of margin is to tolerance the deviation.
+        # We need a larger margin on SPC2/3
+        if self.asic_type != 'spc1':
+            xoff['pkts_num_margin'] = 3
+        self.qos_params_mlnx[self.speed_cable_len]['xoff_1'].update(xoff)
+        self.qos_params_mlnx[self.speed_cable_len]['xoff_2'].update(xoff)
+
+        xon = {}
+        xon['pkts_num_trig_pfc'] = pkts_num_trig_pfc
+        xon['pkts_num_dismiss_pfc'] = pkts_num_dismiss_pfc
+        xon['pkts_num_hysteresis'] = pkts_num_hysteresis
+        if self.asic_type == 'spc2':
+            xon['pkts_num_margin'] = 2
+        elif self.asic_type == 'spc3':
+            xon['pkts_num_margin'] = 0
+        self.qos_params_mlnx['xon_1'].update(xon)
+        self.qos_params_mlnx['xon_2'].update(xon)
+
+        wm_pg_headroom = self.qos_params_mlnx[self.speed_cable_len]['wm_pg_headroom']
+        wm_pg_headroom['pkts_num_trig_pfc'] = pkts_num_trig_pfc
+        wm_pg_headroom['pkts_num_trig_ingr_drp'] = pkts_num_trig_ingr_drp
+        wm_pg_headroom['cell_size'] = self.cell_size
+        if self.asic_type == 'spc1':
+            wm_pg_headroom['pkts_num_margin'] = 1
+        else:
+            wm_pg_headroom['pkts_num_margin'] = 2
+
+        wm_pg_shared_lossless = self.qos_params_mlnx['wm_pg_shared_lossless']
+        wm_pg_shared_lossless['pkts_num_trig_pfc'] = pkts_num_dismiss_pfc
+        wm_pg_shared_lossless['cell_size'] = self.cell_size
+
+        wm_q_shared_lossless = self.qos_params_mlnx[self.speed_cable_len]['wm_q_shared_lossless']
+        wm_q_shared_lossless['pkts_num_trig_ingr_drp'] = pkts_num_trig_ingr_drp
+        wm_q_shared_lossless['cell_size'] = self.cell_size
+
+        lossy_queue = self.qos_params_mlnx['lossy_queue_1']
+        lossy_queue['pkts_num_trig_egr_drp'] = pkts_num_trig_egr_drp - 1
+        lossy_queue['cell_size'] = self.cell_size
```

Comment:
For some of the parameters, I don't see 'self.qos_params_mlnx ' getting updated with the new values

Reply:
Which one is lost?

Reply:
for eg, wm_pg_shared_lossless, lossy_queue. The values in qos.yml are copied to those dict and updated with new values. How are these getting used in the testcases unless you write them back?

Reply:
Yes. The parameters which are not configuration related are copied back from the qos.yml to the dictionary.
The logic in `QosParamMellanox` is like this:
1. Load `qos.yml` into an internal dictionary.
2. Fetch the configuration by using `QosSaiBase`.
3. Generate configuration related parameters and add them to the internal dictionary.
4. The internal dictionary will be returned to `test_qos_sai`, containing both the parameters generated based on configuration and loaded from `qos.yml`.

Reply:
I did see for xon and xoff in lines 99,100,110,111 where after the values were updated they are written back to self.qos_params_mlnx. Don't see that being done for lot of others like wm_pg_shared_lossless, lossy_queue. If we take lossy_queue eg, 'cell size' and 'pkts_num_trig_egr_drp' are updated (lines 130-132) but they are not written back and we return only self.qos_params_mlnx to the testcases

Reply:
For test cases `xon`, `xoff`, and `wm_shared_lossy`, there are at least 2 group of parameters sharing common parameters. 
Taking `xon` as an example, we have `xon_1` and `xon_2`.
- The common parameters are set into a temporary dict `xon` ([line 102 ~ 105)](https://github.com/Azure/sonic-mgmt/pull/1886/files#diff-59cc3af4e9ac202e45277acd476e10e6R102) and then updated to the original dict `self.qos_params_mlnx['xon_1']` and `self.qos_params_mlnx['xon_2']` by using `update` method. ([line 110, 111](https://github.com/Azure/sonic-mgmt/pull/1886/files#diff-59cc3af4e9ac202e45277acd476e10e6R110)).
- The parameters differing between 2 groups are in the qos.yml and will be loaded into dict in the beginning.

For other test cases that have only 1 group of parameters, all the parameters are set into the original dict directly. For example, `wm_pg_headroom`